pkg/build/strategies/sti/postexecutorstep.go

```diff
@@ -211,47 +218,11 @@ func (step *downloadFilesFromBuilderImageStep) execute(ctx *postExecutorStepCont
 }
 
 func (step *downloadFilesFromBuilderImageStep) downloadAndExtractFile(artifactPath, artifactsDir, containerID string) error {
-	glog.V(5).Infof("Downloading file %q", artifactPath)
-
-	fd, err := ioutil.TempFile(artifactsDir, "s2i-runtime-artifact")
+	res, err := downloadAndExtractFileFromContainer(step.docker, step.tar, artifactPath, artifactsDir, containerID)
 	if err != nil {
-		step.builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
-			utilstatus.ReasonFSOperationFailed,
-			utilstatus.ReasonMessageFSOperationFailed,
-		)
-		return fmt.Errorf("could not create temporary file for runtime artifact: %v", err)
-	}
-	defer func() {
-		fd.Close()
-		os.Remove(fd.Name())
-	}()
-
-	if err := step.docker.DownloadFromContainer(artifactPath, fd, containerID); err != nil {
-		step.builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
-			utilstatus.ReasonGenericS2IBuildFailed,
-			utilstatus.ReasonMessageGenericS2iBuildFailed,
-		)
-		return fmt.Errorf("could not download file (%q -> %q) from container %s: %v", artifactPath, fd.Name(), containerID, err)
-	}
-
-	// after writing to the file descriptor we need to rewind pointer to the beginning of the file before next reading
-	if _, err := fd.Seek(0, os.SEEK_SET); err != nil {
-		step.builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
-			utilstatus.ReasonGenericS2IBuildFailed,
-			utilstatus.ReasonMessageGenericS2iBuildFailed,
-		)
-		return fmt.Errorf("could not seek to the beginning of the file %q: %v", fd.Name(), err)
-	}
-
-	if err := step.tar.ExtractTarStream(artifactsDir, fd); err != nil {
-		step.builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
-			utilstatus.ReasonGenericS2IBuildFailed,
-			utilstatus.ReasonMessageGenericS2iBuildFailed,
-		)
-		return fmt.Errorf("could not extract runtime artifact %q into the directory %q: %v", artifactPath, artifactsDir, err)
+		step.builder.result.BuildInfo.FailureReason = res
 	}
-
-	return nil
+	return err
```

Comment:
This should be inside the if statement, then a return nil at the bottom of the function.